### PR TITLE
Build rework - pure babel, no more tsc

### PR DESCRIPTION
**Pull request by BPScott**
### TL;DR

It's a new build output. It looks like this: https://unpkg.com/browse/@shopify/polaris@5.0.0-alpha-newbuild.3/
The esnext build is now build with just babel, rather than running content through typescript then babel

### WHY are these changes introduced?

The current build is pretty convoluted as it has grown over time. It generates standalone js files, standalone css files, an esnext folder and a styles folder. It does a funky thing where it compiles everything using typescript and then runs all that through babel.  This makes the build basically impenetrable to folks who want to modify it. What if it was simpler?

A simpler, more componentised build would allow folks that copy polaris-react for their new component library to more easily rip out the parts they don't need instead of being afraid to touch it.

### WHAT is this pull request doing?

This PR reimagines the build as by centring most of it around a single run of rollup that takes care of the generating index.js, index.es.js and the esnext folder all in one go. This compiles the above by using babel to read from our typescript source files directly instead creating an intermediate build with `tsc`. We also compile our styles directly into css to files in the esnext build to avoid some overhead within consuming apps.

Overall this means:

- The build is a bit easier to understand, and it is easier for copycats to opt out of certain aspects.
- We build assets directly to their final resting place instead of building, then moving them around afterwards
- The can leverage incremental builds for typescript, which drastically shortens the time for repeated builds (**previously we didn't use any caching and repeated builds took ~40s, now repeated builds take ~20s**)
- Bespoke just-for-polaris configs in sewing kit can eventually be removed.
- *Apps using sewing-kit must use at least 0.132.2 to support this PR*. This is a somewhat aggressive overlap period, however the gains made by allowing us to delete 

So how does this work?

`scripts/build.js` is our entry point. It is now a script that runs a handful of other commands in order:

- Run `copy-polaris-tokens` - copies the contents of polaris-tokens into src/styles, so we can do relative imports of that package
- Run `tsc` - performs a types-only build that creates d.ts files in the `types` folder, but does not generate any d.ts files (bonus, we can use incremental builds for this so repeated builds finish in 2 seconds instead of 20 (**for context a full clean build takes ~40s so this cuts repeated build times in half**).
- Run `downlevel-dts` - the tsc command above creates types that work with our TS version (currently 3.8) but what about consumers using older versions? downlevel-dts takes the types generated above and creates versions compatible with TS 3.4, so we use modern TS features without worrying about support for consumers.
- Run `rollup` - This rollup command generates `index.js`, `index.es.js`, `styles.css`, styles.min.css` and the `esnext` folder.
- Run `copyfiles` to store our README docs in the `docs` folder, and our public sass API in the `styles` folder.
- Run build validation tests. These tests are now written in plain JS rather than Jest tests so that we don't have to wait for a build every test cycle (this saves 20s of a full test run).

The Rollup build is the exciting bit of this. It contains two configs - one to generate the standalone files, and one for the esnext folder. They both use babel to read our source files directly and then output content.

The standalone files are functionally the same as before, though it now uses inlined babel helpers for instead of tslib helpers - which means we can remove our dependency on `tslib` and `babel/runtime`.

The esnext folder is generated through rollup using the `preserveModules` which as the name suggests keeps modules around instead of smushing everything into one file. This gives us a folder structure that looks similar to before but a little more optimized (e.g. rollup strips away reexports  component index files). I also take the opportunity to precomple our scss files within the esnext folder, and we now output plain css files.

### To test

- Checkout the previous commit, do a build and generate package contents as a baseline
  `git checkout version-5.0.0 && yarn && yarn run clean && yarn run build && npm pack && tar zxvf shopify-polaris-*.tgz && mv package package-orig`
- Checkout the `build-rework` branch, generate a package contents 
  `git checkout build-rework && yarn && yarn run clean &&  yarn run build && npm pack && tar zxvf shopify-polaris-*.tgz`
- Run a few diffs to compare the old and new contents.
  - `diff -ru package{-orig,}/types` - Type changes are unaffected
  - `diff -ru package{-orig,}/styles` - Style folder is unaffected
  -`diff -ru package{-orig,}/styles.css` - standalone styles are unaffected
 - diff -ru package{-orig,}/index.es.js - standalone JS is slightly different due to having using babel runtime transforms instead of tslib, but otherwise nothing meaningful

The esnext folder is pretty different because rollup removes transitive reexports. Have a look through there, and see that css files are now present - though converted to a "X.processed.scss" filename and they've been processed with scss. The best way to confirm this works is to use `yarn run build-consumer web` to see that this still works when in a consuming app. (Full disclaimer, I think this should be good, I've not actually tested it in a consuming app just yet).

PR for deliveries: https://github.com/Shopify/deliveries/pull/1218
PR for styleguide: https://github.com/Shopify/polaris-styleguide/pull/3463
Unpkg: https://unpkg.com/browse/@shopify/polaris@5.0.0-alpha-newbuild.3/

### CODE REVIEW

**Comment by GoodForOneFare** [comment]
I'll get to this over the weekend!  In the meantime, if any other reviewer, or @Shopify/web-foundation wants to pitch in, I'm eager to see other folks' thoughts 😸

**Comment by BPScott** [comment]
Once #3032 is merged then we can remove the esnext/styles folder and tell all consumers (using sewing-kit or otherwise) to pull the sass api from `styles/_public-api.scss` instead

File: config/rollup/rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  const babelWebPresetOptions = {
+    modules: 'auto',
+    typescript: true,
+    browsers: browserslist,
+  };
+
+  return [
+    replace({
+      '{{POLARIS_VERSION}}': packageJSON.version,
+      '<%= POLARIS_VERSION %>': packageJSON.version,
```
Comment (BPScott) [inline config/rollup/rollup.config.js:35]: Can be removed once #3031 is merged into master and the v5 branch

**Comment by danrosenthal** [comment]
I'm going to remove myself as a reviewer from this, as I won't have time to look into it, and lack the expertise to give it a thorough review.

**Comment by BPScott** [comment]
RIIGHT. After being a proof-of-concept for a while, I've managed to get back to this and give it some love.

This is now ready for review.

File: README.md [hunk 2]
```diff
@@ -76,7 +76,7 @@ If React doesn’t make sense for your application, you can use a CSS-only versi
 ```html
 <link
   rel="stylesheet"
-  href="https://unpkg.com/@shopify/polaris@4.26.1/styles.min.css"
```
[the remark this answers is not in the document]
Comment (BPScott) [reply README.md:79]: We should mention that people will need to update their URLs it in the migration guide / breaking changes list.


README feels a little odd, as it doesn't seem like the place for "here's how it used to be"

File: examples/cdn-styles/README.md [hunk 3]
```diff
@@ -11,7 +11,7 @@ The CDN hosted CSS file needs to be included in your HTML document. Replace the
 ```html
 <link
   rel="stylesheet"
-  href="https://unpkg.com/@shopify/polaris@4.0.0/styles.min.css"
+  href="https://unpkg.com/@shopify/polaris@4.0.0/styles.css"
```
[the remark this answers is not in the document]
Comment (BPScott) [reply examples/cdn-styles/README.md:14]: What part are you refering to?

https://unpkg.com/@shopify/polaris@4.0.0/styles.css currently exists, and will continue to do so.  This example should eventually be updated to use the new path (https://unpkg.com/@shopify/polaris@5.0.0/dist/styles.css) when v5 is released but we can't update these examples to use v5.0.0 until we actually release v5.0.0.

[the remark this answers is not in the document]
Comment (amrocha) [reply examples/cdn-styles/README.md:14]: Maybe I misunderstood when we talked, I thought you mentioned that linking unpkg css wasn't gonna work anymore

File: config/rollup/namespaced-classname.js [hunk 4]
```diff
@@ -2,51 +2,45 @@ const {basename} = require('path');
...
 
-  file[localName] = className;
-  cache.files[filePath] = file;
-  return className;
-};
+    const suffix = includeHash
+      ? `_${stringHash(name).toString(36).substr(0, 5)}`
+      : '';
+
+    return className + suffix;
```
[the remark this answers is not in the document]
Comment (BPScott) [reply config/rollup/namespaced-classname.js:41]: There's a bit of tension here:

- We don't want folks adding extra css styles onto our classnames (as such overrides have no guarantee of stability), so adding a hash to the end of classnames is an attempt to scare people away from that trying that.
- We want folks writing their own html and using our css only to not have to write obtuse classnames

Solution is that we treat regular consumers and sk consumers differently 

- regular (e.g. using our components and create-react-app) and css-only consumers use [styles.css](https://unpkg.com/browse/@shopify/polaris@5.0.0-alpha-newbuild.3/dist/styles.css#LC68) and don't get a hash - so classnames like `.Polaris-Avatar`
- sewing-kit based consumers use [per-component styles in the esnext folder](https://unpkg.com/browse/@shopify/polaris@5.0.0-alpha-newbuild.3/dist/esnext/components/Avatar/Avatar.css) and do get a hash - so classnames like `.Polaris-Avatar_z763p`.



Personally I think supporting css-only consumers is more effort than it's worth - it a crappy development experience and is incredibly painful to update between versions and IMO we shouldn't encourage it, but others still want to keep it as an option - so we end up with this split.

[the remark this answers is not in the document]
Comment (amrocha) [reply config/rollup/namespaced-classname.js:41]: Ah ok, I thought we already did something like this, did we not have hashes anywhere before?

File: examples/cdn-styles/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (BPScott) [reply examples/cdn-styles/README.md:14]: It'll still work as long as you include the version in there and use the appropriate path for version 4 or version 5.

It will break if you don't include a version number:

https://unpkg.com/@shopify/polaris/styles.css works currently (it redirects to https://unpkg.com/@shopify/polaris@4.26.1/styles.css) but once we ship v5 it will break as it will redirect to https://unpkg.com/@shopify/polaris@5.0.0/styles.css which won't exist.

But imo that's a risk you take if you don't specify a major version number.

File: config/rollup/namespaced-classname.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (BPScott) [reply config/rollup/namespaced-classname.js:41]: Before, in our esnext build we shipped [our source scss files](https://unpkg.com/browse/@shopify/polaris@4.26.1/esnext/components/Avatar/Avatar.scss) which an app using SK would then compile using scss and css-modules - as though that scss file was part of the app itself - and then the classname would be only the hash in production (and a readable class in development).

In this PR we take sewing-kit's ideas of "how should I compile scss files" out of the equation entirely - we ship plain css files, and sewing-kit will load those plain css files without any modification (well, possibly some minification, but the classnames aren't manipulated).

That means that we can evolve our idea of what our css build looks like (maybe we switch to dart-sass, or [modular-css](https://m-css.com/)  or [css-blocks](https://css-blocks.com/) etc) without tieing it to "sewing-kit has to use the same build system.

**Comment by BPScott** [comment]
@tsov / @GoodForOneFare: Gonna call for time on this. You're both pretty busy and I don't want that to block this moving on, so I'm gonna merge as-is. 

I'd still love to go over this with the pair of you at some point and get any feedback though.
